### PR TITLE
prov/gni: fix a problem with iov on receive side

### PR DESCRIPTION
**Pull request by hppritcha**
Plug a resource leak (GNI memory registration resources)
when using XPMEM transfer for cases where iov_cnt is > 1
on the receive side.

Fixes ofi-cray/libfabric-cray#1215

Signed-off-by: Howard Pritchard <howardp@lanl.gov>

### CODE REVIEW

File: prov/gni/src/gnix_msg.c [hunk 1]
```diff
@@ -208,6 +208,52 @@ static void __gnix_msg_pack_data_from_iov(uint64_t dest, size_t dest_len,
...
+static inline int __gnix_msg_register_iov(struct gnix_fid_domain *dom,
+					  const struct iovec *iov,
+					  size_t count,
+					  struct gnix_fid_mem_desc **md_vec)
+{
+	int i, ret = FI_SUCCESS;
+	struct fid_mr *auto_mr;
+
+	for (i = 0; i < count; i++) {
+		auto_mr = NULL;
```
Comment (chuckfossen) [inline prov/gni/src/gnix_msg.c]: Why do you NULL this? It's an output parameter to gnix_mr_reg().

File: prov/gni/src/gnix_msg.c [hunk 2]
```diff
@@ -1191,12 +1252,12 @@ static int __gnix_rndzv_iov_req_build(void *arg)
 	gni_ct_get_post_descriptor_t *cur_ct = NULL;
 	void **next_ct = NULL;
 	int head_off, head_len, tail_len;
+	struct fid_mr *auto_mr;
 
 	GNIX_TRACE(FI_LOG_EP_DATA, "\n");
 	/*
 	 * TODO: xpmem intercept here
```
Comment (chuckfossen) [inline prov/gni/src/gnix_msg.c]: Is there an issue open for this TODO?

File: prov/gni/src/gnix_msg.c [hunk 3]
```diff
@@ -3136,77 +3195,59 @@ ssize_t _gnix_recvv(struct gnix_fid_ep *ep, const struct iovec *iov,
...
 					}
+				}
 
-					req->msg.recv_md[i] =
-						container_of(desc[i],
-						struct gnix_fid_mem_desc,
-						mr_fid);
-
+				/*
+				 * TODO: why does this struct store same
```
Comment (chuckfossen) [inline prov/gni/src/gnix_msg.c]: shall we open an issue for this?

**Review by chuckfossen — APPROVED** [review]
Approve with comments

File: prov/gni/src/gnix_msg.c (hunk 2 above)
Comment (hppritcha) [reply prov/gni/src/gnix_msg.c]: no this TODO should be deleted.

File: prov/gni/src/gnix_msg.c (hunk 3 above)
Comment (hppritcha) [reply prov/gni/src/gnix_msg.c]: good idea to keep track of it.

File: prov/gni/src/gnix_msg.c (hunk 1 above)
Comment (hppritcha) [reply prov/gni/src/gnix_msg.c]: Yeah this is probably not necessary within the loop.  Probably setting to NULL in the declaration may reduce likelihood of compiler warnings.  I'll fix this.

File: prov/gni/src/gnix_msg.c (hunk 3 above)
Comment (hppritcha) [reply prov/gni/src/gnix_msg.c]: looks like we may actually use this field at times for unaligned transfers.

File: prov/gni/src/gnix_msg.c [hunk 4]
```diff
@@ -3136,77 +3189,55 @@ ssize_t _gnix_recvv(struct gnix_fid_ep *ep, const struct iovec *iov,
...
 							   " local buffer: %s\n"
 							   , fi_strerror(-ret));
-
-						for (i--; i >= 0; i--) {
-							fi_close(&req->msg.recv_md[i]->mr_fid.fid);
-						}
-
 						goto err;
 					}
+					req->msg.send_flags |= FI_LOCAL_MR;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_msg.c:3213]: Why are we using FI_LOCAL_MR? We don't support it as defined.

Comment (hppritcha) [reply prov/gni/src/gnix_msg.c:3213]: That was in the original code.  We're using that to indicate that the buffer has been registered internally and not by the user.

Comment (sungeunchoi) [reply prov/gni/src/gnix_msg.c:3213]: NM I misread that.  I was thinking it was FI_LOCAL_COMM.

**Pull request merged**